ocaml/xapi/helpers.ml

```diff
@@ -1234,3 +1234,66 @@ let host_supports_hvm ~__context host =
+
+module Task  = struct (* can't place these functions in task helpers due to circular dependencies *)
+  let wait_for ~__context ~tasks =
+    let our_task = Context.get_task_id __context in
+    (** note that in this definition of `classes` we might duplicate our_task
+      * it turns out that filtering out our_task breaks _at least_ `VM.revert`, so best to leave it like this *)
+    let classes = List.map (fun x -> Printf.sprintf "task/%s" (Ref.string_of x)) (our_task::tasks) in
+    let rec process token =
+      TaskHelper.exn_if_cancelling ~__context; (* First check if _we_ have been cancelled *)
+      let statuses = List.filter_map (fun task -> try Some (Db.Task.get_status ~__context ~self:task) with _ -> None) tasks in
+      let unfinished = List.exists (fun state -> state = `pending) statuses in
+      if unfinished
+      then begin
+        let from = call_api_functions ~__context
+            (fun rpc session_id -> Client.Client.Event.from ~rpc ~session_id ~classes ~token ~timeout:30.0) in
+        debug "Using events to wait for tasks: %s" (String.concat "," classes);
+        let from = Event_types.event_from_of_rpc from in
+        process from.Event_types.token
+      end else
+        ()
+    in
+    process ""
+
+  let to_result ~__context ~of_rpc ~name ~t =
+    wait_for ~__context ~tasks:[t];
+    let fail msg = raise Api_errors.(Server_error (internal_error, [Ref.string_of t; Printf.sprintf "%s: %s"name msg])) in
+    let res =
+      match Db.Task.get_status ~__context ~self:t with
+      | `pending                 -> fail "task shouldn't be pending - we just waited on it"
+      | `cancelled | `cancelling -> raise Api_errors.(Server_error (task_cancelled, [Ref.string_of t]))
+      | `failure                 ->
+        begin match Db.Task.get_error_info ~__context ~self:t with
+        | [] | [_]     -> fail "couldn't extract error info from task"
+        | code::params -> raise (Api_errors.Server_error (code, params))
+        end
+      | `success                 ->
+        begin match Db.Task.get_result ~__context ~self:t with
+        | "" -> fail "nothing was written to task result"
+        | s  -> try Xmlrpc.of_string s |> of_rpc
+                with e -> fail (Printf.sprintf "result wasn't placed in task's (ref = %s) result field. error: %s" (Ref.string_of t) (Printexc.to_string e))
+        end
+      in
+    res
+end
+
+let try_internal_async ~__context
+                       ~(debug_tag : string) (* e.g. debug_tag="VM.migrate_send" *)
+                       (internal_async_fn : unit -> API.ref_task)
+                       (sync_fn : unit -> 'a) (* only used when we are talking to a version of the api without InternalAsync *)
+                       (marshaller : Rpc.t -> 'a) =
+  let res = try
+      let t = internal_async_fn () in `new_api t
+    with
+      Api_errors.Server_error(code, params) when code=Api_errors.message_method_unknown -> `old_api
```

Comment:
The exception will escape if then `when` condition is not met. Is this as expected? You would also do:
```
match internal_async_fn () with
| t ->
| exception Api_errors.Server_error(code, params) when code=Api_errors.message_method_unknown ->
| exception other ->
```

Reply:
yes - this is as expected. the reason we catch the `message_method_unknown` is that we want to know when we are contact a version of xapi which doesn't know about internal async calls, in which case we revert to a synchronous call.

i'll add a comment explaining this, as it's not clear